### PR TITLE
only update session in mysql database when required

### PR DESCRIPTION
**Pull request by DanCech**
This is a small update to help address issues with the mysql session store while we re-architect the session system.

It keeps track of whether the session data has been modified, and only issues an update to the database if the data has been modified or the expiry value is more than 60s old.

My testing indicates that this will reduce the number of UPDATE queries sent to the database quite substantially.

### CODE REVIEW

**Review by bergquist — APPROVED** [review]
Tested locally and it works as expected. :+1: 

I don't understand why we didn't think about this before.

File: pkg/services/session/mysql.go [hunk 1]
```diff
@@ -77,13 +83,22 @@ func (s *MysqlStore) ID() string {
 
 // Release releases resource and save data to provider.
 func (s *MysqlStore) Release() error {
+	newExpiry := time.Now().Unix()
+	if !s.dirty && (s.expiry+60) >= newExpiry {
+		// log.Printf("session/mysql: skipping update for session %s", s.sid)
```
Comment (marefr) [inline pkg/services/session/mysql.go]: Remove this or add debug log statement

File: pkg/services/session/mysql.go [hunk 2]
```diff
@@ -77,13 +83,22 @@ func (s *MysqlStore) ID() string {
...
 	if err != nil {
 		return err
 	}
 
 	_, err = s.c.Exec("UPDATE session SET data=?, expiry=? WHERE `key`=?",
-		data, time.Now().Unix(), s.sid)
+		data, newExpiry, s.sid)
+	s.dirty = false
+	s.expiry = newExpiry
+	// log.Printf("session/mysql: updated session %s", s.sid)
```
Comment (marefr) [inline pkg/services/session/mysql.go]: Remove this or add debug log statement

**Review by marefr — COMMENTED** [review]
Very minor comments. A few tests of the `Release()` logic would be great.

**Comment by torkelo** [comment]
looks good, are we confident enough to include this in v5.4.3?

**Comment by DanCech** [comment]
I feel confident, @cinaglia came up with an almost identical patch independently which is encouraging.

In our discussion we did notice that there isn't any checking of expiry in `queryExists` or `Read`, so it's completely dependent on the cleanup job to remove expired sessions.  That's something we should address also, but I won't want to muddy the waters by including it in this commit.

**Comment by bergquist** [comment]
Will we deploy this before the new session implementation? Otherwise, I think we should avoid introducing changes to the session management.

**Comment by torkelo** [comment]
Well this could be included in patch release soon, new session management won’t be in stable until late February

**Comment by DanCech** [comment]
Not sure of the process to cherry-pick it for the patch release

**Comment by bergquist** [comment]
Whoever cuts the patch release will cherry-pick all PR's labeled with `cherry-pick needed`

**Pull request merged**